File: apps/guzo-api/src/main/java/et/guzo/service/MerchantPlatformService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.domain.entity.MerchantApiKey;
import et.guzo.domain.entity.MerchantWebhook;
import et.guzo.domain.entity.Order;
import et.guzo.domain.entity.WebhookDelivery;
import et.guzo.repository.*;
import et.guzo.security.TokenHasher;
import et.guzo.util.IdUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.Instant;
import java.util.*;

@Service
@RequiredArgsConstructor
public class MerchantPlatformService {

    private static final SecureRandom RANDOM = new SecureRandom();

    private final MerchantApiKeyRepository apiKeyRepository;
    private final MerchantWebhookRepository webhookRepository;
    private final WebhookDeliveryRepository deliveryRepository;
    private final OrderRepository orderRepository;
    private final AddressRepository addressRepository;

    @Transactional(readOnly = true)
    public List<Map<String, Object>> listApiKeys(String merchantId) {
        return apiKeyRepository.findByMerchantIdOrderByCreatedAtDesc(merchantId).stream().map(k -> {
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("id", k.getId());
            row.put("label", k.getLabel());
            row.put("keyPrefix", k.getKeyPrefix());
            row.put("isActive", k.isActive());
            row.put("lastUsedAt", k.getLastUsedAt());
            row.put("createdAt", k.getCreatedAt());
            row.put("expiresAt", k.getExpiresAt());
            return row;
        }).toList();
    }

    @Transactional
    public Map<String, String> createApiKey(String merchantId, String label) {
        String prefix = "gz_" + HexFormat.of().formatHex(randomBytes(4));
        String secret = HexFormat.of().formatHex(randomBytes(16));
        String raw = prefix + "." + secret;

        MerchantApiKey key = new MerchantApiKey();
        key.setId(IdUtil.cuid());
        key.setMerchantId(merchantId);
        key.setLabel(label);
        key.setKeyPrefix(prefix);
        key.setKeyHash(TokenHasher.sha256(raw));
        key.setActive(true);
        key.setCreatedAt(Instant.now());
        apiKeyRepository.save(key);

        return Map.of("apiKey", raw, "keyPrefix", prefix, "id", key.getId());
    }

    @Transactional
    public Map<String, Object> revokeApiKey(String merchantId, String keyId) {
        MerchantApiKey key = apiKeyRepository.findByIdAndMerchantId(keyId, merchantId)
            .orElseThrow(() -> ApiException.notFound("API key not found"));
        key.setActive(false);
        apiKeyRepository.save(key);
        return Map.of("id", key.getId(), "label", key.getLabel(), "keyPrefix", key.getKeyPrefix(), "isActive", key.isActive());
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> listWebhooks(String merchantId) {
        return webhookRepository.findByMerchantIdOrderByCreatedAtDesc(merchantId).stream().map(w -> {
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("id", w.getId());
            row.put("url", w.getUrl());
            row.put("events", w.getEvents());
            row.put("isActive", w.isActive());
            row.put("createdAt", w.getCreatedAt());
            return row;
        }).toList();
    }

    @Transactional
    public MerchantWebhook registerWebhook(String merchantId, String url, String secret) {
        if (url == null || url.isBlank()) throw ApiException.badRequest("url is required");
        MerchantWebhook wh = new MerchantWebhook();
        wh.setId("wh_" + HexFormat.of().formatHex(randomBytes(8)));
        wh.setMerchantId(merchantId);
        wh.setUrl(url.trim());
        wh.setSecret(secret != null && !secret.isBlank() ? secret.trim() : HexFormat.of().formatHex(randomBytes(8)));
        wh.setCreatedAt(Instant.now());
        return webhookRepository.save(wh);
    }

    @Transactional
    public void dispatchEvent(String merchantId, String eventType, String payload) {
        List<MerchantWebhook> hooks = webhookRepository.findByMerchantIdAndActiveTrue(merchantId);
        Instant now = Instant.now();
        for (MerchantWebhook hook : hooks) {
            WebhookDelivery d = new WebhookDelivery();
            d.setId(IdUtil.cuid());
            d.setWebhookId(hook.getId());
            d.setEventType(eventType);
            d.setPayload(payload);
            d.setStatus("PENDING");
            d.setAttempts(0);
            d.setCreatedAt(now);
            deliveryRepository.save(d);
        }
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> listCustomers(String merchantId) {
        List<Order> orders = orderRepository.findByMerchantId(merchantId, PageRequest.of(0, 500)).getContent();
        Map<String, Map<String, Object>> map = new LinkedHashMap<>();

        for (Order order : orders) {
            var dropoff = addressRepository.findById(order.getDropoffAddressId()).orElse(null);
            if (dropoff == null) continue;
            String phone = dropoff.getContactPhone() != null && !dropoff.getContactPhone().isBlank()
                ? dropoff.getContactPhone().trim() : dropoff.getLine1();
            Map<String, Object> existing = map.get(phone);
            if (existing != null) {
                existing.put("orderCount", ((Number) existing.get("orderCount")).intValue() + 1);
                if (order.getCreatedAt().toString().compareTo((String) existing.get("lastOrderAt")) > 0) {
                    existing.put("lastOrderAt", order.getCreatedAt().toString());
                }
            } else {
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("contactName", dropoff.getContactName());
                row.put("contactPhone", dropoff.getContactPhone());
                row.put("line1", dropoff.getLine1());
                row.put("city", dropoff.getCity());
                row.put("orderCount", 1);
                row.put("lastOrderAt", order.getCreatedAt().toString());
                map.put(phone, row);
            }
        }

        return map.values().stream()
            .sorted((a, b) -> Integer.compare(
                ((Number) b.get("orderCount")).intValue(),
                ((Number) a.get("orderCount")).intValue()))
            .toList();
    }

    private static byte[] randomBytes(int len) {
        byte[] b = new byte[len];
        RANDOM.nextBytes(b);
        return b;
    }
}

File: apps/server/src/modules/merchant-platform/merchant-platform.service.ts
import crypto from 'node:crypto';
import { prisma } from '@delivery/database';
import { ApiError } from '../../utils/ApiError.js';

export function hashApiKey(raw: string): string {
  return crypto.createHash('sha256').update(raw).digest('hex');
}

function randomHex(bytes: number): string {
  return crypto.randomBytes(bytes).toString('hex');
}

export async function resolveMerchantForUser(userId: string) {
  const merchant = await prisma.merchant.findUnique({ where: { userId } });
  if (!merchant) throw ApiError.badRequest('Authenticated user is not a merchant');
  return merchant;
}

export class MerchantPlatformService {
  async listApiKeys(merchantId: string) {
    return prisma.merchantApiKey.findMany({
      where: { merchantId },
      orderBy: { createdAt: 'desc' },
      select: {
        id: true,
        label: true,
        keyPrefix: true,
        isActive: true,
        lastUsedAt: true,
        createdAt: true,
        expiresAt: true,
      },
    });
  }

  async createApiKey(merchantId: string, label = 'default') {
    const prefix = `gz_${randomHex(4)}`;
    const secret = randomHex(16);
    const raw = `${prefix}.${secret}`;

    const key = await prisma.merchantApiKey.create({
      data: {
        merchantId,
        label,
        keyPrefix: prefix,
        keyHash: hashApiKey(raw),
      },
      select: { id: true, label: true, keyPrefix: true, createdAt: true },
    });

    return { ...key, apiKey: raw };
  }

  async revokeApiKey(merchantId: string, keyId: string) {
    const key = await prisma.merchantApiKey.findFirst({ where: { id: keyId, merchantId } });
    if (!key) throw ApiError.notFound('API key not found');
    return prisma.merchantApiKey.update({
      where: { id: keyId },
      data: { isActive: false },
      select: { id: true, label: true, keyPrefix: true, isActive: true },
    });
  }

  async listWebhooks(merchantId: string) {
    return prisma.merchantWebhook.findMany({
      where: { merchantId },
      orderBy: { createdAt: 'desc' },
      select: { id: true, url: true, events: true, isActive: true, createdAt: true },
    });
  }

  async registerWebhook(merchantId: string, url: string, secret?: string) {
    if (!url?.trim()) throw ApiError.badRequest('url is required');
    const id = `wh_${randomHex(8)}`;
    return prisma.merchantWebhook.create({
      data: {
        id,
        merchantId,
        url: url.trim(),
        secret: secret?.trim() || randomHex(8),
      },
      select: { id: true, url: true, secret: true, events: true, isActive: true, createdAt: true },
    });
  }

  async setWebhookActive(merchantId: string, webhookId: string, isActive: boolean) {
    const hook = await prisma.merchantWebhook.findFirst({ where: { id: webhookId, merchantId } });
    if (!hook) throw ApiError.notFound('Webhook not found');
    return prisma.merchantWebhook.update({
      where: { id: webhookId },
      data: { isActive },
      select: { id: true, url: true, isActive: true },
    });
  }

  async listDeliveries(merchantId: string, limit = 20) {
    return prisma.webhookDelivery.findMany({
      where: { webhook: { merchantId } },
      orderBy: { createdAt: 'desc' },
      take: limit,
      include: { webhook: { select: { url: true } } },
    });
  }

  async listCustomers(merchantId: string) {
    const orders = await prisma.order.findMany({
      where: { merchantId },
      select: {
        id: true,
        orderNumber: true,
        createdAt: true,
        dropoffAddress: {
          select: { contactName: true, contactPhone: true, line1: true, city: true },
        },
      },
      orderBy: { createdAt: 'desc' },
      take: 500,
    });

    const map = new Map<string, {
      contactName: string | null;
      contactPhone: string | null;
      line1: string;
      city: string;
      orderCount: number;
      lastOrderAt: string;
    }>();

    for (const o of orders) {
      const phone = o.dropoffAddress.contactPhone?.trim() || o.dropoffAddress.line1;
      const existing = map.get(phone);
      if (existing) {
        existing.orderCount += 1;
        if (o.createdAt > new Date(existing.lastOrderAt)) existing.lastOrderAt = o.createdAt.toISOString();
      } else {
        map.set(phone, {
          contactName: o.dropoffAddress.contactName,
          contactPhone: o.dropoffAddress.contactPhone,
          line1: o.dropoffAddress.line1,
          city: o.dropoffAddress.city,
          orderCount: 1,
          lastOrderAt: o.createdAt.toISOString(),
        });
      }
    }

    return [...map.values()].sort((a, b) => b.orderCount - a.orderCount);
  }
}

export const merchantPlatformService = new MerchantPlatformService();
